### ISSUE TITLE
Code Chat

### ISSUE THREAD

**Issue by dannasessha** [evanuisse/nth-prime#1]
This is just a 'dummy' issue so we can review the code line by line, but issues are an important part of github workflow as well. We'll want to at least touch on this (the architecture of development on github) but for now this is just to allow up to comment on the lines of code that are there.

In fact, this is not a great way to review code but because there's not a lot of lines we're just going to go with it for now, we might try a different workflow for these kinds of reviews as soon as the next exercism.

**Comment by dannasessha** [comment]
https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L11

I think re-using `n` here could be unclear because of the way the arguments are passed around. `n` is used in the previous function, so the program will compile, but...

**Comment by AloeareV** [comment]
https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L1-L9

This looks pretty straight-forward, only thing here I see is a potential efficiency improvement. 
Say I'm checking if 28 (for example) is prime, and I get to `28 % 14`: 
28/14 is 2, and I've already checked 28 % 2, which would have returned false already. So if I get to 28 % 14, I know it's not gonna be 0, because 28 / 14 is around (well, in this case, exactly) 2, and I've already checked 2. If 28 is prime, I won't get to 14, so if I get there I know it's prime. If you can figure out the point of no return, you could save a lot of time in each call to `is_prime`. @dannasessha I'm really not sure how clear this explanation is, if you want to weigh in?

**Comment by dannasessha** [comment]
https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L13

... here you bind `i` to a value ...

**Comment by dannasessha** [comment]
https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L15

... and here `i` is passed into `is_prime()` as an argument...

`is_prime(i)...`

**Comment by dannasessha** [comment]
https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L1

... and here what was formerly `i` is taken as argument `n` ...

...so now `i` is `n` (in the scope of `is_prime()`) and `n` is still `n` (but a different value) in the scope of `nth()`. This will compile because the language won't be confused by this, but a human reader might.

To make it even more confusing, I believe the actual idea behind the variable `n` is different between `is_prime()` and `nth()`, ie in `is_prime()` we are testing to see if `n` is prime, in `nth()` `n` is the position of a number the sequence of primes.

**Comment by dannasessha** [comment]
> https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L1-L9
> 
> This looks pretty straight-forward, only thing here I see is a potential efficiency improvement.
> Say I'm checking if 28 (for example) is prime, and I get to `28 % 14`:
> 28/14 is 2, and I've already checked 28 % 2, which would have returned false already. So if I get to 28 % 14, I know it's not gonna be 0, because 28 / 14 is around (well, in this case, exactly) 2, and I've already checked 2. If 28 is prime, I won't get to 14, so if I get there I know it's prime. If you can figure out the point of no return, you could save a lot of time in each call to `is_prime`. @dannasessha I'm really not sure how clear this explanation is, if you want to weigh in?

**Comment by AloeareV** [comment]
> https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L1
> 
> ... and here what was formerly `i` is taken as argument `n` ...
> 
> ...so now `i` is `n` (in the scope of `is_prime()`) and `n` is still `n` (but a different value) in the scope of `nth()`. This will compile because the language won't be confused by this, but a human reader might.
> 
> To make it even more confusing, I believe the actual idea behind the variable `n` is different between `is_prime()` and `nth()`, ie in `is_prime()` we are testing to see if `n` is prime, in `nth()` `n` is the position of a number the sequence of primes.

Yeah, the usage of `i` makes sense, to me, but it's pulling enough weight it might want a clearer name, and the multiple `n`s are a bit confusing. The code is small enough I was able to follow it with little difficulty, but using more verbose variable names is good practice to get into.

**Comment by dannasessha** [comment]
> https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L1-L9
> 
> This looks pretty straight-forward, only thing here I see is a potential efficiency improvement.
> Say I'm checking if 28 (for example) is prime, and I get to `28 % 14`:
> 28/14 is 2, and I've already checked 28 % 2, which would have returned false already. So if I get to 28 % 14, I know it's not gonna be 0, because 28 / 14 is around (well, in this case, exactly) 2, and I've already checked 2. If 28 is prime, I won't get to 14, so if I get there I know it's prime. If you can figure out the point of no return, you could save a lot of time in each call to `is_prime`. @dannasessha I'm really not sure how clear this explanation is, if you want to weigh in?

Yes! I think this is pretty well explained, the idea is that there is a point at which, counting up in your "tests to factor" (the `for` loop ranging from `2..(n - 1)`), that you could know you don't need to test any more, because you would be _double _checking.__ Can you figure out what it is?

**Comment by dannasessha** [comment]
> > https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L1
> > 
> > ... and here what was formerly `i` is taken as argument `n` ...
> > ...so now `i` is `n` (in the scope of `is_prime()`) and `n` is still `n` (but a different value) in the scope of `nth()`. This will compile because the language won't be confused by this, but a human reader might.
> > To make it even more confusing, I believe the actual idea behind the variable `n` is different between `is_prime()` and `nth()`, ie in `is_prime()` we are testing to see if `n` is prime, in `nth()` `n` is the position of a number the sequence of primes.
> 
> Yeah, the usage of `i` makes sense, to me, but it's pulling enough weight it might want a clearer name, and the multiple `n`s are a bit confusing. The code is small enough I was able to follow it with little difficulty, but using more verbose variable names is good practice to get into.

I wouldn't mind there being two `n`s as much if they signified the same thing.

**Comment by dannasessha** [comment]
https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L14-L19

Here I was a little confused because the `if` on line 15 is only indented one space, making it a little harder to read. There are tools to auto format, but generally four spaces indentation per scope is standard. Line 15 through line 18 should be adjusted for easy readability.

**Comment by dannasessha** [comment]
https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L2

I was pleasantly surprised by this range! That's a nice way of handling this, very readable.

**Comment by AloeareV** [comment]
https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L14
https://github.com/evanuisse/nth-prime/blob/69b987367990fcca268ec879d991c4717034d44d/src/lib.rs#L20
We talked a bit about the `while` bit requiring a return of `i - 1`, whereas a `loop` with an explicit `if ... { return i; }`  before the `i += 1` would avoid the need to subtract from `i` in the return. I'm in favor or the `while`, personally, as I'm not fond of using `loop` with manual exit clauses when I don't absolutely need to, but there's differing schools of thought there.

**Comment by AloeareV** [comment]
Generally, this looks pretty optimal in terms of simplicity, any changes I'd make would be to improve efficiency at the cost of added complexity. Very nice!

**Comment by dannasessha** [comment]
If you get a chance when this is "loaded into your brain" you could take a look at other people's solutions which might be interesting and/or instructive; I'd say it's definitely worth 30-60 minutes of your time to do this.

Overall, great work, well done! Looking forward to doing some more.
